packages/backend-platform/src/WebsocketService.ts

```diff
@@ -0,0 +1,1348 @@
+  #connectionPromise: Promise<void> | null = null;
+
+  readonly #pendingRequests = new Map<
+    string,
+    {
+      resolve: (value: unknown) => void;
+      reject: (error: Error) => void;
+      timeout: NodeJS.Timeout;
+    }
+  >();
+
+  #lastError: string | null = null;
+
+  #connectedAt: number | null = null;
+
+  // Simplified subscription storage (single flat map)
+  // Key: subscription ID string (e.g., 'sub_abc123def456')
+  // Value: InternalSubscription object with channels, callback and metadata
+  readonly #subscriptions = new Map<string, InternalSubscription>();
+
+  // Channel-based callback storage
+  // Key: channel name (serves as unique identifier)
+  // Value: ChannelCallback configuration
+  readonly #channelCallbacks = new Map<string, ChannelCallback>();
+
+  /**
+   * Creates a new WebSocket service instance
+   *
+   * @param options - Configuration options including messenger
+   */
+  constructor(
+    options: WebSocketServiceOptions & { messenger: WebSocketServiceMessenger },
+  ) {
+    this.#messenger = options.messenger;
+
+    this.#options = {
+      url: options.url,
+      timeout: options.timeout ?? 10000,
+      reconnectDelay: options.reconnectDelay ?? 500,
+      maxReconnectDelay: options.maxReconnectDelay ?? 5000,
+      requestTimeout: options.requestTimeout ?? 30000,
+    };
+
+    // Register action handlers
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:init`,
+      this.init.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:connect`,
+      this.connect.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:disconnect`,
+      this.disconnect.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:sendMessage`,
+      this.sendMessage.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:sendRequest`,
+      this.sendRequest.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:getConnectionInfo`,
+      this.getConnectionInfo.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:getSubscriptionByChannel`,
+      this.getSubscriptionByChannel.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:isChannelSubscribed`,
+      this.isChannelSubscribed.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:addChannelCallback`,
+      this.addChannelCallback.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:removeChannelCallback`,
+      this.removeChannelCallback.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `BackendWebSocketService:getChannelCallbacks`,
+      this.getChannelCallbacks.bind(this),
+    );
+
+    this.init().catch((error) => {
```

Comment:
It looks like we're doing an asynchronous operation in the controller. We don't have a current guideline on this but that's something that we should really avoid if possible, as it leads to unpredictable behavior.

An asynchronous `init` method is something that's become an unofficial pattern in various places, and fortunately it seems that you've implemented this. What are your thoughts on asking consumers to await the `init` method after initializing the WebSocketService instead of automatically calling it in the constructor?